### PR TITLE
Cmake add module

### PR DESCRIPTION
**Pull request by lbudai**
It's a first -naive- implementation.
Please help me to make it final... then I'll modify all the existing modules.
@MrAnno , @Kokan , @furiel

### CODE REVIEW

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by lbudai** [comment]
I've added some examples to show how things will be look like after `add_module` is merged.

**Comment by kira-syslogng** [comment]
Build FAILURE

File: cmake/add_module.cmake [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  cmake_parse_arguments(ADD_MODULE "" "TARGET" "GRAMMAR;SOURCES;DEPENDS;INCLUDES" ${ARGN})
+
+  generate_y_from_ym("modules/${ADD_MODULE_TARGET}/${ADD_MODULE_GRAMMAR}")
+  bison_target(${ADD_MODULE_TARGET}Grammar}
+    ${CMAKE_CURRENT_BINARY_DIR}/${ADD_MODULE_GRAMMAR}.y
+    ${CMAKE_CURRENT_BINARY_DIR}/${ADD_MODULE_GRAMMAR}.c
+    COMPILE_FLAGS ${BISON_FLAGS}
+    )
+
+  include_directories(${ADD_MODULE_INCLUDES})
```
Comment (Kokan) [inline cmake/add_module.cmake]: This is not okay, it should be added to the `target_include_directories`

**Comment by lbudai** [comment]
@kira-syslogng: retest this please

File: modules/afamqp/CMakeLists.txt [hunk 2]
```diff
@@ -1,36 +1,24 @@
 module_switch(ENABLE_AFAMQP "Enable afamqp module" RabbitMQ_FOUND)
 if (ENABLE_AFAMQP)
 
-set(AFAMQP_HEADERS
+set(AFAMQP_SOURCES
```
Comment (Kokan) [inline modules/afamqp/CMakeLists.txt]: Please keep the headers and source file list separet as currently the only use for headers is to install them. (and if you use IDE generator)

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: cmake/add_module.cmake (hunk 1 above)
Comment (lbudai) [reply cmake/add_module.cmake]: Well.. this is another pattern that was accepted until now ;)

File: modules/afamqp/CMakeLists.txt (hunk 2 above)
Comment (lbudai) [reply modules/afamqp/CMakeLists.txt]: Yes, and this separation is also broken in some modules.

**Comment by lbudai** [comment]
@kokan: thanks for the review notes

Unfortunately our CMake build system is a mess...
I hope that finally we can fix it. After this PR is merged, we have to focus on how to improve our review process in order to avoid this situation in the future.

File: modules/afamqp/CMakeLists.txt (hunk 2 above)
Comment (lbudai) [reply modules/afamqp/CMakeLists.txt]: Final conclusion here is to
 * merge private headers with sources (and this is better for IDEs)
 * use `X_HEADERS` for public headers (which we want to be installed with syslog-ng)

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by lbudai** [comment]
let's summarize what happened from latest review notes:
 * `GRAMMAR` is optional as @Kokan suggested (much better in this way)
 * added an example (`basicfuncs`) when `GRAMMAR` is not defined
 * when `GRAMMAR` is set, automatically append the generated grammar files to the source set,
so it is more easier to use at call sites
 * `include_headers()` removed, only `target_include_directories()` is used

I think we are close to the final form of `add_module()`...

@Kokan : could you check it again?

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by Kokan** [comment]
@lbudai LGTM

**Comment by lbudai** [comment]
okay, then I'm going to modify all the modules (this is why I changed status to WIP)

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by lbudai** [comment]
@kira-syslogng : retest this please

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by lbudai** [comment]
okay, it's ready to merge from my part 
(`native`, `openbsd` and `java` where I did not move to `add_module`)

Please check the new `module_generate_y_from_ym`.

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: modules/snmptrapd-parser/CMakeLists.txt [hunk 3]
```diff
@@ -1,37 +1,20 @@
...
-  install(TARGETS snmptrapd-parser
-  LIBRARY DESTINATION lib/syslog-ng/
-  COMPONENT snmptrapd-parser)
+add_module(
+  TARGET snmptrapd-parser
+  GRAMMAR snmptrapd-parser-grammar
+  SOURCES ${SNMPTRAPD_PARSER_SOURCES}
+)
 
-  add_test_subdirectory(tests)
```
[the remark this answers is not in the document]
Comment (lbudai) [reply modules/snmptrapd-parser/CMakeLists.txt]: thanks, done!

File: cmake/Modules/GenerateYFromYm.cmake [hunk 4]
```diff
@@ -39,3 +39,15 @@ function(generate_y_from_ym FileWoExt)
...
 
+# This function is used by add_module
+function(module_generate_y_from_ym FileWoExtSrc FileWoExtDst)
+    if (${ARGC} EQUAL 2)
+      add_custom_command (OUTPUT ${FileWoExtDst}.y
+        COMMAND ${PYTHON_EXECUTABLE} ${PROJECT_SOURCE_DIR}/lib/merge-grammar.py ${FileWoExtSrc}.ym > ${FileWoExtDst}.y
+            DEPENDS ${PROJECT_SOURCE_DIR}/lib/cfg-grammar.y
+            ${FileWoExtSrc}.ym)
+    else()
+        message(SEND_ERROR "Wrong usage of generate_y_from_ym() function")
```
[the remark this answers is not in the document]
Comment (lbudai) [reply cmake/Modules/GenerateYFromYm.cmake]: done

**Comment by lbudai** [comment]
@Kokan : plugin-creator also uses `add_module`, I tried it and it worked (at least for me...).

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: modules/dbparser/CMakeLists.txt [hunk 5]
```diff
@@ -64,27 +64,15 @@ set(DBPARSER_SOURCES
...
-
-add_library(dbparser SHARED ${DBPARSER_SOURCES})
-target_link_libraries(dbparser PRIVATE patterndb)
-target_include_directories(dbparser
-	PUBLIC ${CMAKE_CURRENT_SOURCE_DIR}
-	PRIVATE ${CMAKE_CURRENT_BINARY_DIR}
+add_module(
+  TARGET dbparser
+  GRAMMAR dbparser-grammar
+  INCLUDES ${CMAKE_CURRENT_SOURCE_DIR} ${PATTERNDB_INCLUDE_DIR}
```
Comment (furiel) [inline modules/dbparser/CMakeLists.txt]: Do we need `CMAKE_CURRENT_SOURCE_DIR`? If I understand correctly, `add_module` adds it too.

**Pull request merged**